### DISCUSSION TITLE
What data is stored in various components of standalone Milvus?

### DISCUSSION THREAD

**Discussion by shauryagoel** [milvus-io/milvus#25739] [Q&A and General discussion] [answered]
In standalone Milvus, there are mainly 3 components- milvus, etcd and MinIO. I know that etcd stores metadata of milvus' components, and indexes are stored in MinIO. But, where are the vectors themselves stored? Consequently, where are the entities stored?
Also, what does the Milvus service store? Any other things which are stored in MinIO?

**Comment by yhmo** [comment]
Indexes data, vectors data, and scalar fields(float, int, bool, varchar) data is stored in MinIO.
"entities" is a concept of row-based. Milvus storage is column-based.
For example, a collection has 2 fields: id and vectors.
You insert two entities:
collection.insert({"id": 1, "vectors": [0.1, 0.2, .... 0.5]})
collection.insert({"id": 2, "vectors": [2.0, 2.1, .... 2.5]})

The data of field "id" is stored as a column-based file in MinIO.
The data of filed "vectors" is stored as a column-based file in MinIO.

If you create index for the "vectors" field, the index file is also stored in MinIO.

**Reply by shauryagoel** [reply]
And what does the Milvus service store? Because I am seeing a significant portion of storage space taken by the volumes/milvus directory.

**Reply by yhmo** [reply, marked as answer]
For standalone, there are 3 directories under the "volumes":
- etcd ------ stores metadata
- minio ----- index files, vector data files, scalar data files
- milvus ---- write-ahead-log files, all the data inserted is pre-written in this folder, then the data nodes and query nodes read the data from this folder. WAL is mainly for data safety.
